internal/k8s/controller.go

```diff
@@ -3409,55 +3421,63 @@ func (lbc *LoadBalancerController) getEndpointsForIngressBackend(backend *networ
 		return nil, false, err
 	}
 
-	result, err = lbc.getEndpointsForPort(endps, backend.Service.Port, svc)
+	result, err = lbc.getEndpointsForPortFromEndpointSlices(endpointSlices, backend.Service.Port, svc)
 	if err != nil {
-		glog.V(3).Infof("Error getting endpoints for service %s port %v: %v", svc.Name, configs.GetBackendPortAsString(backend.Service.Port), err)
+		glog.V(3).Infof("Error getting endpoitnslices for service %s port %v: %v", svc.Name, configs.GetBackendPortAsString(backend.Service.Port), err)
 		return nil, false, err
 	}
 	return result, false, nil
 }
 
-func (lbc *LoadBalancerController) getEndpointsForPort(endps api_v1.Endpoints, backendPort networking.ServiceBackendPort, svc *api_v1.Service) ([]podEndpoint, error) {
-	var targetPort int32
-	var err error
-
+func (lbc *LoadBalancerController) validateTargetPort(backendPort networking.ServiceBackendPort, svc *api_v1.Service) (targetPort int32, err error) {
 	for _, port := range svc.Spec.Ports {
 		if (backendPort.Name == "" && port.Port == backendPort.Number) || port.Name == backendPort.Name {
 			targetPort, err = lbc.getTargetPort(port, svc)
 			if err != nil {
-				return nil, fmt.Errorf("error determining target port for port %v in Ingress: %w", backendPort, err)
+				return 0, fmt.Errorf("error determining target port for port %v in Ingress: %w", backendPort, err)
 			}
 			break
 		}
 	}
 
-	if targetPort == 0 {
+	return targetPort, nil
```

Comment:
Do we not need to return an error if targetPort is 0 any more?

Reply:
We do still return an error. It happens within the `getEndpointsForProtFromEndpointSlices` function. I moved code that validated the target port into its own function called `validateTargetPort`. The error we return now will happen if we get an error back from that function.

Reply:
This is the validatePort function 😄.  I see we are returning an error, above on line 3437, but we aren't checking if targetPort is 0 as far as I can see before returning. We could have a case where no targetPort has been found for the backendPort/service combinations but no errors have ben noted. TargetPort will have been initialised to 0, which is not a valid port and will cause errors later in the code where it will be harder to find, so we need to check this somewhere.

Reply:
Ah, I see what you mean. Thanks for catching that. I'll add a check and a unit test for that case.